app/frontend/src/utils/transformUtils.js

```diff
@@ -64,3 +65,147 @@ export function attachAttributesToLinks(formSchemaComponents) {
+          closeDate:newDate.add(keepAliveFor,keepAliveForInterval).format('YYYY-MM-DD HH:MM:SS'),
+          graceDate:newDate.add(allowLateTerm,allowLateInterval).format('YYYY-MM-DD HH:MM:SS')
+        }));
+      }
+      substartDate.add(term,interval);
+    }
+  }
+
+  if((term == null && interval == null) && (keepAliveFor && keepAliveForInterval)){
+    var newDates = substartDate.clone();
+    availableDates.push(Object({
+      startDate:substartDate.format('YYYY-MM-DD HH:MM:SS'),
+      closeDate:newDates.add(keepAliveFor,keepAliveForInterval).format('YYYY-MM-DD HH:MM:SS'),
+      graceDate: allowLateTerm && allowLateInterval ? newDates.add(allowLateTerm,allowLateInterval).format('YYYY-MM-DD HH:MM:SS') : null
+    }));
+  }
+  return availableDates;
+}
+
+
+
+/**
+ * @function getCalculatedCloseSubmissionDate
+ * Get calculated Close date for a Form schedule setting with the given scenario
+ *
+ * @param {Object[]} openDate An object of Moment JS date
+ * keepOpenForTerm
+ * keepOpenForInterval
+ * @param {Integer} term An integer of number of Days/Weeks OR Years
+ * @param {String} interval A string of days,Weeks,months
+ * @param {Integer} allowLateTerm An integer of number of Days/Weeks OR Years
+ * @param {String} allowLateInterval A string of days,Weeks,months
+ * @param {Integer} repeatSubmissionTerm An integer of number of Days/Weeks OR Years
+ * @param {String} repeatSubmissionInterval A string of days,Weeks,months
+ * @param {Object[]} repeatUntil An object of Moment JS date
+ * @param {Object[]} closeDate and object of moment JS date
+ * @returns {Object[]} An object of Moment JS date
+ */
+export function getCalculatedCloseSubmissionDate(openedDate=moment(),keepOpenForTerm=0,keepOpenForInterval='days',allowLateTerm=0,allowLateInterval='days',repeatSubmissionTerm=0,repeatSubmissionInterval='days',repeatSubmissionUntil=moment()){
+  const openDate = moment(openedDate).clone();
+  var calculatedCloseDate = moment(openDate);
+  repeatSubmissionUntil = moment(repeatSubmissionUntil);
+  if(!allowLateTerm && !allowLateInterval && !repeatSubmissionTerm && !repeatSubmissionInterval){
+    calculatedCloseDate = openDate.add(keepOpenForTerm,keepOpenForInterval).format('YYYY-MM-DD HH:MM:SS');
+  }else{
+    if(repeatSubmissionTerm && repeatSubmissionInterval && repeatSubmissionUntil){
+      calculatedCloseDate = repeatSubmissionUntil;
+    }
+    if(allowLateTerm && allowLateInterval){
+      calculatedCloseDate = calculatedCloseDate.add(keepOpenForTerm,keepOpenForInterval).add(allowLateTerm,allowLateInterval).format('YYYY-MM-DD HH:MM:SS');
+    }
+  }
+  return calculatedCloseDate;
+}
+
+
+/**
+ * @function calculateCloseDate
+ * Get close date when provided a given period for late submission
+ *
+ * @param {Integer} allowLateTerm An integer of number of Days/Weeks OR Years
+ * @param {String} allowLateInterval A string of days,Weeks,months
+ */
+export function calculateCloseDate(
+  subcloseDate=moment(),
```

Comment:
What is this? Why isn't it in the function documentation? The descriptions need to be clear